File: Body/S/S4/ta-onta/shared/coordinate_phase.ts
export type CoordinatePhase = "direct" | "prime" | "inverted_property";

export interface PhaseQualifiedCoordinate {
  raw: string;
  family: string;
  position: string;
  phase: CoordinatePhase;
  handle: string;
}

export interface PhasePreservationQuery {
  query: "preserve_coordinate_phase";
  surface: string;
  source_coordinate: string;
  emitted_coordinate: string;
  required_phase: CoordinatePhase;
}

export function resolvePhaseQualifiedCoordinate(coord: string): PhaseQualifiedCoordinate {
  const raw = coord.trim();
  if (!raw) throw new Error("coordinate is required");
  const property = parsePropertyCoordinate(raw);
  if (property) return property;

  const family = raw[0];
  if (!family || !["C", "P", "L", "S", "T", "M"].includes(family)) {
    throw new Error(`unsupported coordinate family ${family ?? ""}`);
  }
  const rest = raw.slice(1);
  const phase: CoordinatePhase = rest.startsWith("'") || rest.endsWith("'") ? "prime" : "direct";
  const position = rest.startsWith("'")
    ? rest.slice(1)
    : rest.endsWith("'")
      ? rest.slice(0, -1)
      : rest;
  return {
    raw,
    family,
    position,
    phase,
    handle: `vak://${phase}/${raw}`,
  };
}

export function phasePreservationQuery(
  sourceCoordinate: string,
  emittedCoordinate: string,
  surface = "full-7-laws",
): PhasePreservationQuery | null {
  const source = resolvePhaseQualifiedCoordinate(sourceCoordinate);
  const emitted = resolvePhaseQualifiedCoordinate(emittedCoordinate);
  if (
    source.family === emitted.family &&
    source.position === emitted.position &&
    source.phase !== "direct" &&
    emitted.phase === "direct"
  ) {
    return {
      query: "preserve_coordinate_phase",
      surface,
      source_coordinate: source.raw,
      emitted_coordinate: emitted.raw,
      required_phase: source.phase,
    };
  }
  return null;
}

export function phaseQualifiedVakToken(method: string, coord: string): string {
  const resolved = resolvePhaseQualifiedCoordinate(coord);
  return `<vak: method="${method}" coord="${resolved.raw}" phase="${resolved.phase}" handle="${method}(${resolved.raw})">`;
}

function parsePropertyCoordinate(raw: string): PhaseQualifiedCoordinate | null {
  const parts = raw.split("_");
  if (parts.length < 3) return null;
  const [family, position, marker] = parts;
  if (!family || family.length !== 1 || !/^[a-zA-Z]$/.test(family)) return null;
  if (!position || !/^\d+$/.test(position)) return null;
  const phase: CoordinatePhase = marker === "i" ? "inverted_property" : "direct";
  return {
    raw,
    family,
    position,
    phase,
    handle: `vak://${phase}/${raw}`,
  };
}

File: Body/S/S0/portal-core/src/coordinate_phase.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct PhaseQualifiedCoordinate {
    pub raw: String,
    pub family: String,
    pub position: String,
    pub phase: String,
    pub handle: String,
}

pub fn resolve_phase_qualified_coordinate(coord: &str) -> Result<PhaseQualifiedCoordinate, String> {
    let raw = coord.trim();
    if raw.is_empty() {
        return Err("coordinate is required".to_owned());
    }

    if let Some(parsed) = parse_property_coordinate(raw) {
        return Ok(parsed);
    }

    let mut chars = raw.chars();
    let family = chars
        .next()
        .ok_or_else(|| "coordinate is required".to_owned())?;
    if !matches!(family, 'C' | 'P' | 'L' | 'S' | 'T' | 'M') {
        return Err(format!("unsupported coordinate family `{family}`"));
    }

    let rest: String = chars.collect();
    let (position, phase) = if let Some(stripped) = rest.strip_prefix('\'') {
        (stripped.to_owned(), "prime")
    } else if let Some(stripped) = rest.strip_suffix('\'') {
        (stripped.to_owned(), "prime")
    } else {
        (rest, "direct")
    };

    Ok(PhaseQualifiedCoordinate {
        raw: raw.to_owned(),
        family: family.to_string(),
        position,
        phase: phase.to_owned(),
        handle: format!("vak://{phase}/{raw}"),
    })
}

pub fn assert_coordinate_phase_preserved(source: &str, emitted: &str) -> Result<(), String> {
    let source = resolve_phase_qualified_coordinate(source)?;
    let emitted = resolve_phase_qualified_coordinate(emitted)?;
    if source.family == emitted.family
        && source.position == emitted.position
        && source.phase != "direct"
        && emitted.phase == "direct"
    {
        return Err(format!(
            "phase-erasing coordinate emission: {} ({}) collapsed to {} ({})",
            source.raw, source.phase, emitted.raw, emitted.phase
        ));
    }
    Ok(())
}

pub fn phase_qualified_vak_token(method: &str, coord: &str) -> Result<String, String> {
    let resolved = resolve_phase_qualified_coordinate(coord)?;
    Ok(format!(
        "<vak: method=\"{}\" coord=\"{}\" phase=\"{}\" handle=\"{}({})\">",
        method, resolved.raw, resolved.phase, method, resolved.raw
    ))
}

fn parse_property_coordinate(raw: &str) -> Option<PhaseQualifiedCoordinate> {
    let parts: Vec<&str> = raw.split('_').collect();
    if parts.len() < 3 {
        return None;
    }
    let family = parts[0];
    let position = parts[1];
    if family.len() != 1 || !family.chars().all(|c| c.is_ascii_alphabetic()) {
        return None;
    }
    if !position.chars().all(|c| c.is_ascii_digit()) {
        return None;
    }
    let phase = if parts.get(2) == Some(&"i") {
        "inverted_property"
    } else {
        "direct"
    };
    Some(PhaseQualifiedCoordinate {
        raw: raw.to_owned(),
        family: family.to_owned(),
        position: position.to_owned(),
        phase: phase.to_owned(),
        handle: format!("vak://{phase}/{raw}"),
    })
}
